Fix bad interaction of anchor positioning and overscroll

fantasai opened the pull request:
#### 43c5d324c08a2c8f6c5ba88a499435f50975b73e
<pre>
Fix bad interaction of anchor positioning and overscroll
<a href="https://bugs.webkit.org/show_bug.cgi?id=298031">https://bugs.webkit.org/show_bug.cgi?id=298031</a>
<a href="https://rdar.apple.com/159356009">rdar://159356009</a>

Reviewed by Antti Koivisto.

This patch fixes several problems that are triggered during overscroll:
1. We need to clamp the scroll position to avoid laying out into the
   overscroll area.
2. We need to not recapture scroll positions when we are running layout
   only on the children (and therefore not recalculating the rect itself).
3. We need to ensure that a scroll-dependent anchor() gets recalculated
   whenever layout triggers a recapture of the scroll positions, so that
   the capture and the anchor() resolution don&apos;t get out of sync.

* LayoutTests/fast/scrolling/anchor-overscroll-fixed-expected.html: Added.
* LayoutTests/fast/scrolling/anchor-overscroll-fixed.html: Added.

Add test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt: Removed.

Pass more tests.

* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::registerAnchorScrollAdjuster):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:

Return whether or not the recapture differs.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::simplifiedLayout):

Don&apos;t capture during layout that skips rect calculation, so that the
rect and the capture stay in sync.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::constrainedScrollPosition const):

Add helper method to get the constrained scroll position.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::AnchorScrollAdjuster::recaptureDiffers const):

Add helper method to identify differing captures.

(WebCore::AnchorScrollAdjuster::addSnapshot):
(WebCore::AnchorScrollAdjuster::addViewportSnapshot):
(WebCore::AnchorScrollAdjuster::adjustmentForViewport const):

Constrain the scroll position.

(WebCore::Style::AnchorPositionEvaluator::captureScrollSnapshots):

Allow calling captureScrollSnapshots() from style resolution.
Trigger style invalidation when the capture differs.

(WebCore::Style::offsetFromAncestorContainer):
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):

Constrain the scroll position.

(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Capture scroll positions when resolving anchor() functions so they can
be compared to the positions captured during layout.

* Source/WebCore/style/AnchorPositionEvaluator.h:

Update API for above changes.

Canonical link: <a href="https://commits.webkit.org/300061@main">https://commits.webkit.org/300061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/772a86a74f96d2e9a497083c5e872360a53e9ccd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121193 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40889 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31547 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127620 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73274 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/691bd9f2-8789-4a35-aab1-999956cfc9f7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123069 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41591 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49468 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92063 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a7137a53-6b01-43f3-9975-38a08a203708) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124145 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108627 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72741 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a0db822f-e1ea-4ffe-9389-ff1a90170a92) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32234 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26734 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71206 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102708 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130465 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48120 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36575 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100666 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48488 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104799 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100570 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25487 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45972 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24029 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44814 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47978 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53691 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47449 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50796 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49133 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -129,15 +135,24 @@ RenderBox* AnchorScrollAdjuster::anchored() const
 {
-    auto position = renderView.frameView().scrollPositionRespectingCustomFixedPosition();
+    auto& view = renderView.frameView();
+    auto position = view.constrainedScrollPosition(IntPoint(view.scrollPositionRespectingCustomFixedPosition()));
```
`IntPoint` -> `ScrollPosition`

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -147,8 +162,9 @@ LayoutSize AnchorScrollAdjuster::adjustmentForViewport(const RenderView& renderV
+        auto& view = renderView.frameView();
         return m_scrollSnapshots.first().m_scrollSnapshot
-            - renderView.frameView().scrollPositionRespectingCustomFixedPosition();
+            - view.constrainedScrollPosition(IntPoint(view.scrollPositionRespectingCustomFixedPosition()));
```
Ditto

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f5ceb4e0fb1da30327c78e3432641760c97b8dad. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50498

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -65,6 +65,12 @@ AnchorScrollSnapshot::AnchorScrollSnapshot(LayoutPoint snapshot)
     , m_scrollSnapshot(snapshot)
 { }
 
+bool AnchorScrollSnapshot::operator==(const AnchorScrollSnapshot& other) const
+{
+    return m_scroller.get() == other.m_scroller.get()
+        && m_scrollSnapshot == other.m_scrollSnapshot;
+}
+
```
`bool AnchorScrollSnapshot::operator==(const AnchorScrollSnapshot&) const = default;` might work here

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7a1cc5d33fe7ae458cb228ecf5ad82914c7ac59a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50498

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/772a86a74f96d2e9a497083c5e872360a53e9ccd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50498

fantasai committed:
Fix bad interaction of anchor positioning and overscroll
https://bugs.webkit.org/show_bug.cgi?id=298031
rdar://159356009

Reviewed by Antti Koivisto.

This patch fixes several problems that are triggered during overscroll:
1. We need to clamp the scroll position to avoid laying out into the
   overscroll area.
2. We need to not recapture scroll positions when we are running layout
   only on the children (and therefore not recalculating the rect itself).
3. We need to ensure that a scroll-dependent anchor() gets recalculated
   whenever layout triggers a recapture of the scroll positions, so that
   the capture and the anchor() resolution don't get out of sync.

* LayoutTests/fast/scrolling/anchor-overscroll-fixed-expected.html: Added.
* LayoutTests/fast/scrolling/anchor-overscroll-fixed.html: Added.

Add test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt: Removed.

Pass more tests.

* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::registerAnchorScrollAdjuster):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:

Return whether or not the recapture differs.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::simplifiedLayout):

Don't capture during layout that skips rect calculation, so that the
rect and the capture stay in sync.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::constrainedScrollPosition const):

Add helper method to get the constrained scroll position.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::AnchorScrollAdjuster::recaptureDiffers const):

Add helper method to identify differing captures.

(WebCore::AnchorScrollAdjuster::addSnapshot):
(WebCore::AnchorScrollAdjuster::addViewportSnapshot):
(WebCore::AnchorScrollAdjuster::adjustmentForViewport const):

Constrain the scroll position.

(WebCore::Style::AnchorPositionEvaluator::captureScrollSnapshots):

Allow calling captureScrollSnapshots() from style resolution.
Trigger style invalidation when the capture differs.

(WebCore::Style::offsetFromAncestorContainer):
(WebCore::Style::AnchorPositionEvaluator::computeAnchorRectRelativeToContainingBlock):

Constrain the scroll position.

(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Capture scroll positions when resolving anchor() functions so they can
be compared to the positions captured during layout.

* Source/WebCore/style/AnchorPositionEvaluator.h:

Update API for above changes.

Canonical link: https://commits.webkit.org/300061@main
